classes/plugins/PluginHelper.php

```diff
@@ -43,51 +44,63 @@ class PluginHelper
      * @param string $filePath Full path to plugin archive
      * @param string $originalFileName Original filename of plugin archive
      *
-     * @return string Extracted plugin path
+     * @return string Directory where the plugin was extracted
      */
-    public function extractPlugin($filePath, $originalFileName)
+    private function extractPlugin(string $filePath, string $originalFileName): string
     {
+        // A permissive extension might be returned (e.g. 1.2.3.tar.gz)
+        $getExtension = fn (string $path) => explode('.', basename($path), 2)[1] ?? '';
+        // Drops risky characters
+        $sanitize = fn (string $path) => preg_replace('/[^\w.-]/', '', $path);
+
+        $extension = $sanitize($getExtension($originalFileName));
+        $baseName = $sanitize(basename($originalFileName, ".{$extension}")) ?: 'plugin';
         $fileManager = new FileManager();
-        // tar archive basename (less potential version number) must
-        // equal plugin directory name and plugin files must be in a
-        // directory named after the plug-in (potentially with version)
-        $matches = [];
-        PKPString::regexp_match_get('/^[a-zA-Z0-9]+/', basename($originalFileName, '.tar.gz'), $matches);
-        $pluginShortName = array_pop($matches);
-        if (!$pluginShortName) {
-            throw new Exception(__('manager.plugins.invalidPluginArchive'));
-        }
 
-        // Create random dirname to avoid symlink attacks.
-        $pluginExtractDir = dirname($filePath) . "/{$pluginShortName}" . substr(md5(mt_rand()), 0, 10);
-        if (!mkdir($pluginExtractDir)) {
-            throw new Exception('Could not create directory ' . $pluginExtractDir);
+        // If the extension doesn't match, copy the file to another location to avoid issues with the PharData class
+        $filePathWithExtension = null;
+        if ($getExtension($filePath) !== $extension) {
```

Comment:
This is a fatal error, I think?

Reply:
Why exactly? FYI these updates are currently deployed at the OJS 3.4 test installation, the basic operations are working fine:
- Installation
- Installation from plugin gallery
- Removal
- Upgrade
- Upgrade from plugin gallery
- Error conditions (e.g. upgrade an already upgraded plugin)

Reply:
Ah, the use of lambda functions looked at a glance like a typo.

We already have `FileManager::getExtension`; I'd suggest making sure that handles compound extensions, and working with it instead.

Otherwise I'd suggest just making a private or protected function in the class rather than using lambdas this way.

Reply:
Yep, I just wanted to check if it's ok to keep the `FileManager` alive, I'll take a look if I can move some things to there and make use of the `Flysystem`.

---

About the `FileManager::getExtension()` it won't work with `.tar.gz` (also its implementation needs to be fixed: `getExtension('test')` returns "test"). The `FileManager::parseFileExtension()` works (it has an exception for this extension), but I decided to have something more generic 🤔